frontend/src/metabase/lib/redux/utils.js

```diff
@@ -270,45 +271,55 @@ function withCachedData(
   // thunk decorator:
   return thunkCreator =>
     // thunk creator:
-    (...args) =>
-    // thunk:
-    (dispatch, getState) => {
-      const options = args[args.length - 1] || {};
-      const { useCachedForbiddenError, reload, properties } = options;
-
-      const existingStatePath = getExistingStatePath(...args);
-      const requestStatePath = ["requests", ...getRequestStatePath(...args)];
-      const newQueryKey = getQueryKey && getQueryKey(...args);
-      const existingData = getIn(getState(), existingStatePath);
-      const { loading, loaded, queryKey, error } =
-        getIn(getState(), requestStatePath) || {};
-
-      // Avoid requesting data with permanently forbidded access
-      if (useCachedForbiddenError && error?.status === 403) {
-        throw error;
-      }
+    (...args) => {
+      async function thunk(dispatch, getState) {
+        const options = args[args.length - 1] || {};
+        const { useCachedForbiddenError, reload, properties } = options;
+
+        const existingStatePath = getExistingStatePath(...args);
+        const requestStatePath = ["requests", ...getRequestStatePath(...args)];
+        const newQueryKey = getQueryKey && getQueryKey(...args);
+        const existingData = getIn(getState(), existingStatePath);
+        const { loading, loaded, queryKey, error } =
+          getIn(getState(), requestStatePath) || {};
+
+        // Avoid requesting data with permanently forbidded access
+        if (useCachedForbiddenError && error?.status === 403) {
+          throw error;
+        }
+
+        const hasRequestedProperties =
+          properties &&
+          existingData &&
+          _.all(properties, p => existingData[p] !== undefined);
+
+        // return existing data if
+        if (
+          // we don't want to reload
+          // the check is a workaround for EntityListLoader passing reload function to children
+          reload !== true &&
+          // reload if the query used to load an entity has changed even if it's already loaded
+          newQueryKey === queryKey
+        ) {
+          // and we have a non-error request state or have a list of properties that all exist on the object
+          if (loaded || hasRequestedProperties) {
+            return existingData;
+          } else if (loading) {
+            // wait for current loading request to be resolved
+            const result = await waitFor(
```

Comment:
I'd rather follow our pattern with storing promises for the original request. Something like this https://github.com/metabase/metabase/blob/ca2059f34e505093b0c0987307a70f763e1c032d/frontend/src/metabase/query_builder/actions/querying.js#L267. What do you think? Alternatively we can store promises outside of redux.